src/frontend/core/executor/impl/TriangleCountExecutor.cpp

```diff
@@ -604,109 +965,93 @@ int TriangleCountExecutor::updateTriangleTreeAndGetTriangleCount(std::vector<std
+            string workerId = partitionWorkerMap[part];
+            auto workerWeightMapIterator = workerWeightMap.find(workerId);
             if (workerWeightMapIterator != workerWeightMap.end()) {
-                int weight = workerWeightMap.at(workerId);
+                int weight = workerWeightMapIterator->second;
 
                 if (minimumWeight == 0 || minimumWeight > weight) {
                     minimumWeight = weight + 1;
                     minWeightWorker = workerId;
+                    minWeightWorkerPartition = part;
                 }
             } else {
                 minimumWeight = 1;
                 minWeightWorker = workerId;
+                minWeightWorkerPartition = part;
             }
         }
+        workerWeightMap[minWeightWorker] = minimumWeight;
 
-        string aggregatorSqlStatement =
-            "SELECT ip,user,server_port,server_data_port,partition_idpartition "
-            "FROM worker_has_partition INNER JOIN worker ON worker_has_partition.worker_idworker=worker.idworker "
-            "WHERE partition_graph_idgraph=" +
-            graphId + " and idworker=" + minWeightWorker + ";";
-
-        std::vector<vector<pair<string, string>>> result = sqlite->runSelect(aggregatorSqlStatement);
-
-        vector<pair<string, string>> aggregatorData = result.at(0);
-
-        std::string aggregatorIp = aggregatorData.at(0).second;
-        std::string aggregatorUser = aggregatorData.at(1).second;
-        std::string aggregatorPort = aggregatorData.at(2).second;
-        std::string aggregatorDataPort = aggregatorData.at(3).second;
-        std::string aggregatorPartitionId = aggregatorData.at(4).second;
+        const auto &workerData = workerDataMap[minWeightWorker];
+        std::string aggregatorIp = workerData[0];
+        std::string aggregatorPort = workerData[1];
+        std::string aggregatorDataPort = workerData[2];
 
-        if ((aggregatorIp.find("localhost") != std::string::npos) || aggregatorIp == masterIP) {
-            aggregatorHost = aggregatorIp;
-        } else {
-            aggregatorHost = aggregatorUser + "@" + aggregatorIp;
-        }
+        std::string aggregatorPartitionId = minWeightWorkerPartition;
 
-        for (aggregatorCopyCombinationIterator = workerCombination.begin();
-             aggregatorCopyCombinationIterator != workerCombination.end(); ++aggregatorCopyCombinationIterator) {
-            std::string workerId = *aggregatorCopyCombinationIterator;
-            string host = "";
+        std::string partitionIdList = "";
+        for (auto partitionCombinationIterator = partitionCombination.begin();
+             partitionCombinationIterator != partitionCombination.end(); ++partitionCombinationIterator) {
+            string part = *partitionCombinationIterator;
+            string workerId = partitionWorkerMap[part];
 
+            if (part != minWeightWorkerPartition) {
+                partitionIdList += part + ",";
```

Comment:
Can we avoid string concatenation and use a string builder instead?

Reply:
It will run at most 3 times, so the overhead is negligible

Reply:
Ok